Add review checklist (#895)
zeitgeistpm/zeitgeist: docs/review_checklist.md

- [ ] Check if there are any TODOs which are not linked to an issue of the
      zeitgeist repository like this: `TODO(#999)`.
- [ ] Does the PR link relevant issues and contain a detailed description?
- [ ] If the PR changes the business logic, does it add the corresponding
      labels?
- [ ] Are the doc strings up to date?
- [ ] If the PR adds or changes extrinsics or functions used by extrinsics:
  - [ ] Is the _Weight_ section in the documentation up to date?
  - [ ] Are the benchmarks up to date?
  - [ ] Do the call filters need to be adjusted (forbidden scoring rules,
        dispute mechanisms)?
  - [ ] Do the extrinsics emit all the required events (see [Events](#events)
        below)?
- [ ] Is the module `README.md` up to date?
- [ ] Is [docs.zeitgeist.pm] up to date?
- [ ] Is `docs/changelog_for_devs.md` up to date, specifically:
  - [ ] Are changes relevant to the Frontend Team (extrinsics changed, new
        functions) mentioned here?
  - [ ] Are all new events explained so they can easily be integrated into the
        indexer?
  - [ ] Are breaking changes marked as such?
  - [ ] Is the file formatted with `prettier -w docs/changelog_for_devs.md`?
- Sanity tests:

>   - [ ] The local node produces blocks.

Author: 14b8bfc250664077b782b9e66d133825a9372add

  - [ ] `try-runtime` passes on Zeitgeist and Battery Station.
  - [ ] All runtime benchmarks pass on Zeitgeist and Battery Station (don't just

>         test against the mock!).

Author: 98ee686664182f188383546f6d0f093de65df68b

>   - [ ] The node syncs with Zeitgeist and Battery Station.

Author: Not exactly. By "local node" I mean the local temporary dev node. Checking that the node syncs with the live networks is a different thing. I've added commands for each in 14b8bfc250664077b782b9e66d133825a9372add.

- [ ] Code quality:

>   - [ ] Are there any compiler/clippy warnings?

Author: edd3b1c17321a1a92727bf92694c5a8c4dc4a4c4

>   - [ ] Are there any compiler/clippy warnings?

Author: (Not all compiler warnings seem to be caught by the CI.)

>   - [ ] Is integer arithmetic saturated/checked and are all panickers removed?

Author: Feel free to raise an issue about this. Might be interesting to have this, although I think this is a difficult thing to solve. For example, will these tools still allow us to use substrate calls, which often contain panickers?

  - [ ] Is code contained in macro invocations (benchmarks,
        `runtime/common/src/lib.rs`, `decl_runtime_apis!`) correctly formatted?

>   - [ ] Are all `*.toml` files formatted with `taplo`?

Author: 98ee686664182f188383546f6d0f093de65df68b

  - [ ] Is the _Weight_ section in the documentation up to date?
  - [ ] Are the benchmarks up to date?
  - [ ] Do the call filters need to be adjusted (forbidden scoring rules,
        dispute mechanisms)?
  - [ ] Do the extrinsics emit all the required events (see [Events](#events)
        below)?
- [ ] Is the module `README.md` up to date?
- [ ] Is [docs.zeitgeist.pm] up to date?
- [ ] Is `docs/changelog_for_devs.md` up to date, specifically:
  - [ ] Are changes relevant to the Frontend Team (extrinsics changed, new
        functions) mentioned here?
  - [ ] Are all new events explained so they can easily be integrated into the
        indexer?
  - [ ] Are breaking changes marked as such?
  - [ ] Is the file formatted with `prettier -w docs/changelog_for_devs.md`?
- Sanity tests:
  - [ ] The local temporary development node produces blocks:
        `cargo run --profile=production -- --tmp`.
  - [ ] The node syncs with Zeitgeist and Battery Station:
        `cargo run --profile=production --features=parachain`,
        `cargo run --profile=production --features=parachain -- --chain=battery_station`.
  - [ ] `try-runtime` passes on Zeitgeist and Battery Station.

> - [ ] Code quality:
>   - [ ] Are there compiler warnings which can easily be avoided?
>   - [ ] Is integer arithmetic saturated/checked and are all panickers removed?
>   - [ ] Is code contained in macro invocations (benchmarks,
>         `runtime/common/src/lib.rs`, `decl_runtime_apis!`) correctly formatted?
>   - [ ] Are all `*.toml` files formatted with `taplo` (run
>         `taplo format --check`)?
>   - [ ] Are all copyright notices up to date?
>   - [ ] Are enums sorted alphabetically, except for enums used in storage (to
>         prevent migrations) or errors and events?

Reviewer: ```suggestion
- [ ] Code quality:
  - [ ] Avoidable compiler warnings were resolved.
  - [ ] Integer arithmetic is only saturated/checked and all panickers are removed.
  - [ ] Code contained in macro invocations (benchmarks,
        `runtime/common/src/lib.rs`, `decl_runtime_apis!`) is correctly formatted.
  - [ ] All `*.toml` files are formatted with `taplo` (run
        `taplo format --check`).
  - [ ] All copyright notices are up to date.
  - [ ] Enums are sorted alphabetically, except for enums used in storage (to
        prevent migrations), errors and events.
```

- [ ] If an action is required by the Frontend Team, add an issue to
      zeitgeistpm/ui.
- [ ] If the PR adds a new pallet, is the pallet added to the benchmark
      configuration in `scripts/`?

> - [ ] If you are changing/removing configuration values, storage items: Do the
>       changes require a storage migration?

Reviewer: ```suggestion
- [ ] In case configuration items or storage elements were changed: Necessary storage migrations are provided.
```

> - [ ] If you are adding/changing configuration values on the mainnet: Have the
>       implications been discussed with the product owners?

Reviewer: Adjusted two things here beside the conversion to an action:
1. Removed the constraint "mainnet only", because changing values can also lead to undesired issues on the testnet
2. Replaced project owners by code owners.

```suggestion
- [ ] In case configuration values changed: The
      implications have been discussed with the [code owners](https://github.com/zeitgeistpm/zeitgeist/blob/main/CODEOWNERS).
```

> - [ ] If the changes include a storage migration:
>   - [ ] Is the version number in the pallet bumped?
>   - [ ] Add try-runtime checks and ensure the following:
>     - [ ] The storage migration bumps the pallet version correctly.
>     - [ ] The try-runtime _fails_ if you comment out the migration code.
>     - [ ] The try-runtime passes without any warnings (substrate storage
>           operations often just log a warning instead of failing, so these
>           warnings usually point to problem which could break the storage).

Reviewer: ```suggestion
- [ ] If the changes include a storage migration:
  - [ ] The affected pallet's `STORAGE_VERSION` was bumped.
  - [ ] Try-runtime checks were added and the following conditions were ensured:
    - [ ] The storage migration bumps the pallet version correctly.
    - [ ] The try-runtime _fails_ if you comment out the migration code.
    - [ ] The try-runtime passes without any warnings (substrate storage
          operations often just log a warning instead of failing, so these
          warnings usually point to problem which could break the storage).
```

## Events

> _All_ modifications of the on-chain storage **must** be broadcast to our indexer

Reviewer: ```suggestion
_All_ modifications of the on-chain storage **must** be broadcast to indexers
```

by emitting a high-level event. The term _high-level_ event refers to an event
which may or may not contextualize _low-level_ events emitted by pallets that

> our business logic builds on, for example pallet-balances. Examples of

Reviewer: ```suggestion
Zeitgeist's business logic builds on, for example pallet-balances. Examples of
```
